DM-31725: rewrite queries subpackage, via new dependency on daf_relation

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Integrate relations with QueryBuilder and dimension managers.

Most of this ticket is a major overhaul of QueryBuilder and the
DimensionRecordStorage classes to use relations throughout, with
conversion back to the old SimpleQuery system now happening only at
the boundary between QueryBuilder and Query.

Ideally, that'd be plenty for one commit, but doing that work revealed
some preexisting problems with our dimension caching system and how we
test for data ID validity in a few different contexts.  But I couldn't
fix those in earlier commits without the relation machinery introduced
on this one, and I couldn't get that new relation machinery to pass
tests without fixing them.  So here they are.

The first problem was that all the previous cleverness about
populating the CachingDimensionRecordStorage cache one row at a time
never made sense: it's much more efficient to just fetch all rows for
cached dimensions whenever any of them are requested, since by
definition we only configure caching for dimensions whose record count
is small enough that we can easily hold them all in memory, and
probably small enough that fetching all of them is still dominated by
per-query overheads.

Changing that fixes a bug in which QueryDimensionRecordStorage (used
exclusively for 'band') happily returned nonexistent dimension records
if you gave it a nonexistent data ID, and hence you could expand a bad
data ID and not notice.  Or, more precisely, this commit sidesteps
that still-existing problem, because 'band' is actually configured to
use CachingDimensionRecordStorage on top of
QueryDimensionRecordStorage, and now CachingDimensionRecordStorage
prevents the bad behavior.  In a few commits I plan to remove the
fetch method from all of those storage objects, and then the bug will
really fully be gone.

Anyhow, fixing that bug creates new problems, in the form of a
Registry test that was actually checking for the wrong behavior - we
*can* tell when a query with a bad band value is doomed, because we
try to expand the data ID up front and (now) that fails.  But we don't
want an exception for this - we want a doomed query.  So that
snowballed into additional logic changes involving exceptions
vs. doomed queries.

Last but not least, this modifies CachingDimensionRecordStorage and
GovernorDimensionRecordStorage to allow them to have an unset (None)
cache, and removes the refresh methods for populating those caches.
Calling higher-level refresh methods now actually unsets the caches,
forcing them to be repopulated on first use.  It didn't actually make
sense to have both refresh and clearCaches before, but we need
clearCaches in order to have something we can call at transaction
rollback without hitting the database; this is exactly the problem
identified (with collection caching) on DM-35498.  Also, having
clearCaches make the cache an empty container was a bug waiting to
happen for GovernorDimensionRecordStorage, since that (usually
incorrectly) would say that there are no governor dimension values in
the database.

TallJimbo committed:
Add column tag classes and implement daf_relation interfaces.

This adds a dependency on daf_relation and specializations of a few of
its types for daf_butler.  This doesn't actually do anything yet.

TallJimbo committed:
Add row readers for data IDs, DatasetRefs, and dimension records.

These are not yet used anywhere, and they partially duplicate
functionality in the Query.extract* methods, which will soon be
replaced by these new readers.

Add relation-row readers for DimensionRecords and expanded data IDs.

TallJimbo committed:
Add factories for common predicates.

These are not used yet, but they will be soon (and we'll be adding
more, too, as needed).

TallJimbo committed:
Add QueryContext and QueryBackend objects.

These will eventually form the backbone of the new query system; for
now they're unused stubs.

TallJimbo committed:
Add ColumnTypeInfo.make_relation_table_spec.

TallJimbo committed:
Integrate relations into dataset subqueries and storage managers.

This is the first of several commits that replace pieces of Registry
query code with new versions that are based on the new daf_relation
package.  The real value of that switch won't be realized until the
query system as a whole is replaced (and in particular the Query class
itself is wholly rewritten), and in the meantime there will be a fair
amount of transitional code tying the new and old interfaces together.

This moves the former DatasetRecordStorage.find implementation into
SqlRegistry, since it can now delegate to
DatasetRecordStorage.make_relation.  And once we have a QueryBackend
implementation for RemoteRegistry, we can move it further into the
base class Registry and use the same implementation for all
registries.

TallJimbo committed:
Rewrite query WHERE-clause handling via relation predicates.

This gives the queries.expressions package a single entry point, which
is annoying in its unit tests' imports but I think a very nice win for
code organization overall, in addition to setting up this aspect of
the query system for the future.

The commit also includes a few minor (but broad) style changes:
renaming dataId->data_id arguments in some internal interfaces, and
using just the empty string for nonexistent expressions instead of
allowing either None or the empty string (at the level of type
annotations, anyway).

It also includes one small behavior change: instead of check=False
turning off all expression analysis (which disabled some important
checks, and made it hard to reason about the behavior in their
absence), it now just permits expressions to leave dimensions
"orphaned" by not providing their governor.  In internal interfaces
the argument has been renamed to allow_orphans to reflect this.

Finally, this still includes some transitional code that will
ultimately be removed when Relation and Predicate are more fully
integrated with the query system.  We still have two versions
(i.e. new and old) of many many things.

Restore range literals in expressions have inclusive stop.

Needs a test.

TallJimbo committed:
Use relations to rewrite Query and QueryResults.

This is a full rewrite of Query - it's now a concrete final class
instead a hierarchy (extensibility is provided by the QueryContext and
QueryBackend interfaces it delegates to), and it's all backed by
Relation instead of interacting with SQLAlchemy objects at all.

That's important for future client-server butler work, where the
expectation is that Query can be used more-or-less as-is with new
QueryContext and QueryBackend implementation (with the caveat that an
abstract interface rarely survives the definition of its second-ever
derived class unscathed).

It's also where Relation really starts to show its power in making the
query system _conceptually_ simpler, even if there's still a ton of
code: before, when applying a new operation (e.g. sorting) to an
existing query, we basically had to try to reconstruct all of the
parameters that went into the original query, then modify them
according to the new operation (and any that had been applied since).
Effectively every Query factory method had to know about all of the
things that could go into construting a Query.  Now, we can sometimes
just append a new relation operation to the existing tree, and even
the more complex cases can be expressed as relatively straightforward
operations on that existing tree.

That gives us the backbone we'll need for real query composability;
it's not quite here yet only because I'm focusing now on
reimplementing existing interfaces rather than defining newer more
powerful ones.  But it's a pretty small jump from here to a lot of
things we've wanted for a long time, especially QG generation,
including:

- data ID set uploads, implemented as iteration-engine leaf relations that
  get transferred to the SQL engine;

- adding new dimensions to an existing query with customization of the
  spatiotemporal relationships to use, which is just a frontend to
  QueryBackend.make_dimension_relation;

- vectorizing calibration and other prerequisite lookups, which is a
  little more code in `Query.find_datasets` to add temporal joins.
  And the previously-thorny problem how how to represent the results
  of that - which need to involve rows that represent both a
  DatasetRef and dimensions that aren't part of that ref's DatasetType
  - is neatly solved by direct iteration over a Query instead of
  DatasetQueryResults.

On that note, the other big change here is that pretty much all of the
functionality in the QueryResults classes has moved to Query; the
QueryResults classes are just thin facades now, and
DimensionRecordQueryResults now delegates directly to Query (as its
data ID and dataset siblings do) instead of the query-factory stuff it
had before.

The changes in tests here mostly reflect the fact that using relation
improves our ability to do diagnostics on queries that return no
results as well as our ability to see when a query is doomed before
execution.  I'm a bit worried the diagnostics in some cases could also
be getting a bit less readable to end-users with this change, though,
because they're coming from daf_relation now and use its terminology.
But I think they'll still carry the important pieces of information
and they'll be much more precise, so let's see how they work in the
wild before trying to intercept and rewrite them in daf_butler.

TallJimbo committed:
Pass view target storage instance to QueryDimensionRecordStorage.

This simplifies its join implementation by allowing it to delegate to
the target's join method.  It'll also help with some new methods in
future commits.

TallJimbo committed:
Temporarily drop exception message tests for record order-by.

These exception messages reflect parsing an order-by str sequence in
the context of a single DimensionElement, but logic that will land
shortly instead parses these in the context of multiple elements (i.e.
a data ID query).  I'll restore these tests and the slightly better
error messages they represent after the daf_relation overhaul is
complete.

TallJimbo committed:
Add DimensionElementFields.columns property.

TallJimbo committed:
Revert "Temporarily drop exception message tests for record order-by."

This reverts commit 771d589ccab86d211bd2d967d286430d8f90998c.

TallJimbo committed:
Drop DimensionRecordStorage.fetch.

All previous uses went away with the Query rewrite.

TallJimbo committed:
Drop DataCoordinateIterable.constrain.

This was only used by DimensionRecordStorage.fetch and
DatasetRecordStorage.[de]certify, and both of those have now been
switched to use relations instead.

TallJimbo committed:
Add make_data_id_relation to QueryContext.

TallJimbo committed:
Use relation in certify/decertify implementations.

This requires passing a SqlQueryContext down from the Registry itself
through a few layers.  That'll be increasingly common as we move more
read operations to rely on relation-returning methods to reduce the
SQL-aware interface surface.

TallJimbo committed:
Drop SimpleQuery.

All usage now fully replaced by relation.

TallJimbo committed:
Drop QueryColumns and DatasetQueryColumns.

TallJimbo committed:
Prohibit caches of views of dimension records.

This avoids a problem with dimension record cache consistency on was
introduced in the cache rework a few commits back, when we started
considering the cache more trustworthy.

Unfortunately there's no cost-free fix: that cache rework fixed other
bugs with cache consistency, so we can't just drop it.  The best
solution seems to be admitting that expandDataIds cannot be _relied_
upon to test for dimension record existence; this has always been the
case in general, but it may have been impossible to trigger via the
default dimension configuration before.

See code comments for additional rationale.

TallJimbo committed:
Note that findDataset now respects a given storage class.

TallJimbo committed:
Add changelog entry.

TallJimbo committed:
Add daf_relation dependency to pyproject.toml.

timj commented:
To confirm what I said on Jira: `lsst-daf-relation` will have to be added as a dependency in `pyproject.toml` once that package ends up on PyPI. (which I see happened just after I wrote this)

andy-slac commented on python/lsst/daf/butler/core/_column_type_info.py:
```diff
@@ -46,10 +59,92 @@ class ColumnTypeInfo:
+                            nullable=False,
+                        )
+                    )
+                case DimensionRecordColumnTag(element=element_name, column="region"):
```
`element_name` does not seem to be used in this `case`?

andy-slac commented on python/lsst/daf/butler/registry/queries/_readers.py:
```diff
@@ -0,0 +1,336 @@
+
+    Parameters
+    ----------
+    full_reader ` : `_FullDataCoordinateReader`
```
Extra backtick.

andy-slac commented on python/lsst/daf/butler/registry/queries/_readers.py:
```diff
@@ -0,0 +1,336 @@
+    record_caches : `Mapping`
+        Nested mapping (outer keys are dimension elements, inner keys are data
+        IDs for that element) of cached dimension records.  Ignored unless
+        ``records=True``.
```
Last sentence probably not needed here.

andy-slac commented on python/lsst/daf/butler/registry/queries/_readers.py:
```diff
@@ -0,0 +1,336 @@
+        self._record_readers = record_readers
+        self._record_caches = record_caches
+
+    __slots__ = ("_full_reader", "_record_readers", "_record_caches", "_cached_columns_required")
```
`_cached_columns_required` is not used?

TallJimbo replied:
It's used by the `@cached_getter` decorator.

andy-slac commented on python/lsst/daf/butler/registry/queries/_readers.py:
```diff
@@ -0,0 +1,336 @@
+        "_translate_collection",
+        "_id_tag",
+        "_run_tag",
+        "_cached_columns_required",
```
This one too?

TallJimbo replied:
It's used by `@cached_getter`.

andy-slac commented on python/lsst/daf/butler/registry/queries/_readers.py:
```diff
@@ -0,0 +1,336 @@
+        run : `str`, optional
+            Name of the `~CollectionType.RUN` collection; when provided the run
+            key does not need to be present in the result row, and
+            ``translate_collection` does not need to be provided at
```
Needs one more closing backtick.

andy-slac commented on python/lsst/daf/butler/registries/sql.py:
```diff
@@ -439,38 +440,104 @@ def findDataset(
+        if len(components) > 1:
+            raise DatasetTypeError(
+                f"findDataset requires exactly one dataset type; got multiple components {components} "
+                f"for parent dataset type{parent_dataset_type.name}."
```
Space needed after `dataset type`?

andy-slac commented on python/lsst/daf/butler/registries/sql.py:
```diff
@@ -439,38 +440,104 @@ def findDataset(
+            collections = self.defaults.collections
+        backend = queries.SqlQueryBackend(self._db, self._managers)
+        collection_wildcard = CollectionWildcard.from_expression(collections)
+        collection_wildcard.require_ordered()
```
This feels a bit fragile, maybe we want `require_ordered=True` argument to `from_expression()`?

andy-slac commented on python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py:
```diff
@@ -360,188 +329,189 @@ def select(
+            used to construct the new relation.
+        requested_columns : `~collections.abc.Set` [ `str` ]
+            Columns the relation should include.
+        collections : `Sequence` [ `CollectionRecord` ]
```
Should it be 

    `Sequence` [`tuple`[`CollectionRecord`, `int`]]

andy-slac commented on python/lsst/daf/butler/registry/queries/_query_backend.py:
```diff
@@ -213,3 +265,470 @@ def resolve_single_dataset_type_wildcard(
+        ----------
+        dataset_type : `DatasetType`
+            Dataset type to be queried in the returned collections.
+        collections : `CollectionSearch` or `CollectionWildcard`
```
Type annotation only mentions `CollectionWildcard`.

andy-slac commented on python/lsst/daf/butler/registry/queries/_structs.py:
```diff
@@ -22,208 +22,143 @@
-        Mapping containing literal values that should be injected into the
-        query expression, keyed by the identifiers they replace.
+    Instances of this class should only be created by
+    `QueryWhereExpression.attach`, which guarantees the consistency of its
```
`QueryWhereClause.combine` instead?

andy-slac commented on python/lsst/daf/butler/registry/queries/_structs.py:
```diff
@@ -22,208 +22,143 @@
-    expression.
+    expression_predicate: Predicate | None
+    """A predicate that evaluates a string expression from the user
+    (`ExpressionPredicate` or `None`).
```
Predicate instead of ExpressionPredicate?

andy-slac commented on python/lsst/daf/butler/registry/queries/_structs.py:
```diff
@@ -345,14 +344,18 @@ class QuerySummary:
+    data_id : `DataCoordinate`, optional
         A fully-expanded data ID identifying dimensions known in advance.  If
         not provided, will be set to an empty data ID.
     expression : `str` or `QueryWhereExpression`, optional
```
Should be "str, optional"

andy-slac commented on python/lsst/daf/butler/registry/queries/expressions/_predicate.py:
```diff
@@ -0,0 +1,460 @@
+
+    def visitRangeLiteral(self, start: int, stop: int, stride: int | None, node: Node) -> VisitorResult:
+        # Docstring inherited.
+        return ColumnContainer.range_literal(range(start, stop, stride or 1))
```
In our language both start and stop are inclusive, is that handled correctly in the code that converts ranges to SQL (Python `range` looks suspicious and I'm not sure that we have a unit test for it)?

TallJimbo replied:
Good catch!  I hadn't realized our `stop` here was inclusive; I'll add a `+1` here as well as a unit test.

andy-slac commented on python/lsst/daf/butler/registry/queries/expressions/check.py:
```diff
@@ -83,6 +87,33 @@ def update(self, other: InspectionSummary) -> None:
+        if self.hasIngestDate:
+            if dataset_type_name is None:
+                raise UserExpressionError(
+                    "Expression requires an ingest data, which requires " "exactly one dataset type."
```
data -> date? `" "` in the middle can be removed.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_dimensions.py:
```diff
@@ -91,45 +95,36 @@ def clearCaches(self) -> None:
+            dimension element and assume that dataset or spatial join relations
+            that might provide these will be included in the relation tree
+            first.
+        join : `~lsst.daf.relation.Join`
+            Join operation to use when the implementation is an actual join.
+            When a true join is being simulated by other relation operations,
+            this objects `~lsst.daf.relation.Join.min_columns` and
+            `~lsst.daf.relation.Join.max_columns` should still be respected.
```
Not sure why `join` is needed, aren't we supposed to use `Relation.join` for joins?

TallJimbo replied:
I've defined the interface this way to allow the skypix dimension storage to simulate joins by actually applying a column-calculation relation operation that adds the region columns.  That behaves like a join in this case because we know the virtual skypix table has to have a superset of the skypix indices appearing in any other table, and there are never any other columns in those virtual tables besides index and the region.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_dimensions.py:
```diff
@@ -202,34 +197,92 @@ def sync(self, record: DimensionRecord, update: bool = False) -> bool | dict[str
-        tables : `Iterable` [ `sqlalchemy.schema.Table` ]
-            Possibly empty set of tables for schema digest calculations.
+        relation : `lsst.daf.relation.Relation`
+            New relation.
```
Annotation says it's still returning iterable of `Table`.

TallJimbo replied:
I think this must have been copy-pasted into the wrong place, but looking a little closer, I think the annotation should really be `list[sqlalchemy.schema.Table]`, since we do `+=` on it and that's not legal on general iterables.  Interesting that mypy didn't complain about that; I wonder if it can't do arithmetic operators for some reason.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_dimensions.py:
```diff
@@ -273,32 +326,54 @@ def element(self) -> GovernorDimension:
-        `refresh` to ensure up-to-date results.
+        Returns
+        -------
+        relations : `~lsst.daf.relation.Relation`
```
`relation` instead of `relations`?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_dimensions.py:
```diff
@@ -273,32 +326,54 @@ def element(self) -> GovernorDimension:
+        (`sqlalchemy.schema.Table`).
         """
-        raise NotImplementedError()
+        raise NotImplementedError
```
Parentheses after NotImplementedError?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_dimensions.py:
```diff
@@ -372,6 +451,39 @@ def element(self) -> DatabaseDimensionElement:
+
+        Returns
+        -------
+        relations : `~lsst.daf.relation.Relation`
```
relations again

andy-slac commented on python/lsst/daf/butler/registry/dimensions/governor.py:
```diff
@@ -164,21 +154,27 @@ def sync(self, record: DimensionRecord, update: bool = False) -> bool | dict[str
+                record = reader.read(row)
+                cache[record.dataId] = record
+            self._cache = cache
+        return cast(Mapping[DataCoordinate, DimensionRecord], self._cache)
```
I think there is a big overlap in cache management between this module and `caching`, maybe factor that out in some mixing class (in a future ticket)?

TallJimbo replied:
:+1: I want to revisit some of the caching logic at a higher level, too, to deal with problems with cache expiration on rollback like [DM-35498](https://jira.lsstcorp.org/browse/DM-35498), since I think that's broader than just the  collections issue it mentions and I think the approach taken here of allowing caches to be empty is the way out.  I'm repurposing that ticket to expand its scope to this now.

andy-slac commented on python/lsst/daf/butler/registry/dimensions/skypix.py:
```diff
@@ -62,41 +64,42 @@ def clearCaches(self) -> None:
 
     def insert(self, *records: DimensionRecord, replace: bool = False, skip_existing: bool = False) -> None:
-        # Docstring inherited from DimensionRecordStorage.insert.
+        # Docstring inherited from DimensionRecordStorage.insert.W
```
Cat **W**alked on keyboard? 😸

andy-slac commented on python/lsst/daf/butler/registry/obscore/_manager.py:
```diff
@@ -66,58 +61,44 @@ class _ExposureRegionFactory(ExposureRegionFactory):
-
-        detector = cast(Dimension, self.universe["detector"])
-        if detector in dataId:
+        relation = visit_definition_storage.join(relation, Join(), context)
```
Is joining with a special "initial" relation results in just a regular relation for visit_definition_storage? Could not you use `visit_definition_storage.make_relation()` here? (Just trying to get a hang at how it all works together).

TallJimbo replied:
That would be totally fine at runtime.  But for MyPy I'd also need to do `cast(DatabaseDimensionRecordStorage, visit_definition_storage)` since `DimensionRecordStorage` doesn't have `make_relation`.

andy-slac commented on python/lsst/daf/butler/registry/queries/_builder.py:
```diff
@@ -49,80 +50,27 @@ class QueryBuilder:
-        executed.  Queries with a non-empty list will never be executed.
+    context : `.queries.SqlQueryContext`
+        Object that manages relation engines and database-side state
+        (e.g. temporary tables) for the query.
```
Type annotation says it's `QueryContext` (or `None`). Don't context and backend always come together, or you can use context from some other backend here?

TallJimbo replied:
Docs got out of date; in some of the intermediate commits this did have to be a `SqlQueryContext`, and it looks like I updated the annotation after that but not the docs.

andy-slac commented on python/lsst/daf/butler/registry/queries/_sql_query_backend.py:
```diff
@@ -87,6 +103,225 @@ def resolve_dataset_type_wildcard(
+            raise ColumnError(
+                "Cannot add dimension key column(s) "
+                f"{{{', '.join(name for name in missing_columns.dimension_keys)}}} "
+                f" that were not included in the given dimensions {dimensions}-."
```
Is dash needed after `{dimensions}`? There is a space at the end of previous line and at the start of current one.

andy-slac commented on python/lsst/daf/butler/registry/queries/_sql_query_backend.py:
```diff
@@ -87,6 +103,225 @@ def resolve_dataset_type_wildcard(
+                f" that were not included in the given dimensions {dimensions}-."
+            )
+        if missing_columns.datasets:
+            raise ColumnError("Unexpected dataset columns in call to make_dimension_relation.")
```
Could this message provide more context about what happens?

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -20,1114 +20,935 @@
+    "QueryResults" classes that are backed by it directly exposed to users.  It
+    has been designed with the intent that it will eventually play a larger
+    role, either as the main query result object in a redesigned query
+    interface interface, or a "power user" result option that accompanies
```
`interface` x 2

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -20,1114 +20,935 @@
-            `True` if the query would (or might, depending on arguments) yield
-            result rows.  `False` if it definitely would not.
+        refs : `~collections.abc.Iterator` [ `DatasetRef` ]
+            Iterator that yields dataset references.
```
Will these be resolved or unresolved refs?

TallJimbo replied:
Always resolved.

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -20,1114 +20,935 @@
-        Messages related to post-query filtering are only available if `rows`,
-        `any`, or `count` was already called with the same region (with
-        ``exact=True`` for the latter two).
+        data_ids : `~collections.abc.Iterator` [ `DimensionRecord` ]
```
Maybe `records` instead of `data_ids`?

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -42,1113 +38,917 @@
+        operations (e.g. `with_only_column`, `sliced`, `sorted`) via Python
+        code acting on the existing container rather than going back to SQL,
+        which can be less efficient overall that applying operations to a
+        deferred queryand executing them all only at the end.
```
`queryand`, space missing

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -42,1113 +38,917 @@
+        Running a query is represented in terms of relations by adding a
+        `~lsst.daf.relation.Materialization` marker relation in the iteration
+        engine and then processing the relation tree; this attaches the
+        containre of rows to that new relation to short-circuit any future
```
typo: containre

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -42,1113 +38,917 @@
+
+        Notes
+        -----
+        Adding dimension record columns to is fundamentally different from
```
"to" not needed here?

andy-slac commented on python/lsst/daf/butler/registry/queries/_query.py:
```diff
@@ -42,1113 +38,917 @@
+        defer : `bool`, optional
+            If `False`, run the new query immediately.  If `True`, do not.  If
+            `None` (default), the ``defer`` option passed when making ``self``
+            is used (this option is "sticky").
```
There is no default for `defer`

andy-slac commented on python/lsst/daf/butler/registry/queries/_query_context.py:
```diff
@@ -0,0 +1,463 @@
+        -------
+        modified : `Relation`
+            Stripped relation tree, with engine != `iteration_engine`.
+        stripped : `UnaryOperation`
```
`list`? Does it mean there could no binary operations with iteration engine?

TallJimbo replied:
Yes.  It was a lot simpler this way and we didn't need them - at present the iteration engine doesn't implement joins at all, and while it does implement chains, `daf_butler` doesn't use them.

andy-slac commented on python/lsst/daf/butler/registry/queries/_sql_query_context.py:
```diff
@@ -0,0 +1,522 @@
+                "must be executed in Python in order to obtain an exact "
+                "count.  Pass discard=True to run the query and discard the "
+                "result rows while counting them, run the query first, or "
+                "pass exact=False to obtain a lower bound."
```
lower bound -> upper bound?

andy-slac reviewed (approved):
Looks good, but this is all very new to me, so there are probably some dumb questions. I hoped that we would have less code in `daf_butler` but apparently line count cannot go down. 🙂

TallJimbo committed:
Fix type annotation for digestTables.

We were using += on the results of these calls, which only works for
lists or tuples, not general iterables (and it was always lists at
runtime).

Not sure why MyPy didn't complain; maybe it doesn't check arithmetic
operators?

TallJimbo reviewed (commented):
I believe I've addressed all review comments but two:
- reworking `CollectionWildcard` to make `require_ordered` less fragile;
- adding a test for inclusive vs. exclusive bounds in range literals.

But I'm going to rebase on `main` and resolve some conflicts before I start in on those.

TallJimbo committed:
Add require_ordered kwarg to CollectionWildcard.from_expression.
